cider/tests/utils/CiderBatchChecker.cpp

```diff
@@ -70,35 +70,58 @@ int CiderBatchChecker::getTotalNumOfRows(
 
 bool CiderBatchChecker::checkValidityBitmapEqual(const CiderBatch* expected_batch,
                                                  const CiderBatch* actual_batch) {
-  auto expected_buffer = expected_batch->getNulls();
-  auto actual_buffer = actual_batch->getNulls();
-  auto expected_count = expected_batch->getNullCount();
-  auto actual_count = actual_batch->getNullCount();
+  auto expected_null_cnt = expected_batch->getNullCount();
+  auto actual_null_cnt = actual_batch->getNullCount();
   auto row_num = expected_batch->getLength();
 
   // skip checking null_count if it is not yet computed (-1)
-  if (expected_count != -1 && actual_count != -1 && expected_count != actual_count) {
+  if (expected_null_cnt != -1 && actual_null_cnt != -1 &&
+      expected_null_cnt != actual_null_cnt) {
     std::cout << "expected_null_count != actual_null_count. "
-              << "Expected: " << expected_count << ". Actual: " << actual_count
+              << "Expected: " << expected_null_cnt << ". Actual: " << actual_null_cnt
               << std::endl;
     return false;
   }
 
+  auto expected_buffer = expected_batch->getNulls();
+  auto actual_buffer = actual_batch->getNulls();
+
   if (!expected_buffer && !actual_buffer) {
     // both buffers are nullptr, no need to check
     return true;
   } else if (expected_buffer && actual_buffer) {
     // both buffers exist
     return CiderBitUtils::CheckBitVectorEq(expected_buffer, actual_buffer, row_num);
+  } else {
+    // one is nullptr but the other is not, null_count of the other batch must be 0
+    if (!actual_null_cnt) {
+      // actual values are all valid, expected null count should be 0
+      expected_null_cnt = expected_null_cnt == -1
```

Comment:
1. default null_count is `-1`?
2. should be `CiderBitUtils::countSetBits(null_buffer)` ?

Reply:
1. `null_count` may be `-1` if it is not yet computed ([Apache Arrow doc](https://arrow.apache.org/docs/format/CDataInterface.html#c.ArrowArray.null_count)). But it seems in the current arrow array builder computing `null_count` has not yet been implemented (marked as a TODO), and in current implementation the initial uncomputed value for `null_count` is `0`.
2. Actually an `ArrowArray` does not have a "null buffer". It only has a "validity bitmap", where `1` represents non-null and `0` represents null. It is the inverse of a null buffer, so we should count unset bits (0's, i.e. null values) in the validity bitmap. (The name `getNulls()` is actually a bit misleading because it actually gets a validity bitmap instead of a null bitmap).